src/Type/Php/DatePeriodConstructorReturnTypeExtension.php

```diff
@@ -0,0 +1,54 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use DatePeriod;
+use DateTimeInterface;
+use PhpParser\Node\Expr\StaticCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\DynamicStaticMethodReturnTypeExtension;
+use PHPStan\Type\Generic\GenericObjectType;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\NullType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+class DatePeriodConstructorReturnTypeExtension implements DynamicStaticMethodReturnTypeExtension
+{
+
+	public function getClass(): string
+	{
+		return DatePeriod::class;
+	}
+
+	public function isStaticMethodSupported(MethodReflection $methodReflection): bool
+	{
+		return $methodReflection->getName() === '__construct';
+	}
+
+	public function getTypeFromStaticMethodCall(MethodReflection $methodReflection, StaticCall $methodCall, Scope $scope): Type
+	{
+		$thirdArgType = null;
+		if (isset($methodCall->getArgs()[2])) {
+			$thirdArgType = $scope->getType($methodCall->getArgs()[2]->value);
+		}
+
+		if (
+				$thirdArgType instanceof ObjectType
+			&& $thirdArgType->isInstanceOf(DateTimeInterface::class)->yes()
+		) {
+
+			return new GenericObjectType(DatePeriod::class, [
+				new ObjectType(DateTimeInterface::class),
+				new NullType(),
+			]);
+		}
+
+		return new GenericObjectType(DatePeriod::class, [
```

Comment:
This logic isn't sufficient. You also need to test scenarios like when the third argument is `mixed` or `DateTimeInterface|int`.

Reply:
I made a more explicit version